### PR TITLE
Join React Test completed

### PR DESCRIPTION
**Pull request by JulianKosach**
Live demo is [here](https://join-react-test.herokuapp.com/).

All requirements are implemented.

1. [Home Screen](https://join-react-test.herokuapp.com/)

Screenshot 2020-08-25 at 4 24 15 PM

> On Home Screen user will be able to navigate either to [Candidate View](https://join-react-test.herokuapp.com/applicant) or [Recruiter View](https://join-react-test.herokuapp.com/candidates)

2. [Candidate View](https://join-react-test.herokuapp.com/applicant)

Screenshot 2020-08-25 at 4 24 23 PM

> On Candidate Screen user will be able to apply himself. All Fields are optional there. Also user will be able to upload his photo.

3. [Recruiter View](https://join-react-test.herokuapp.com/candidates)

Screenshot 2020-08-25 at 4 24 30 PM

> On Recruiter Screen user will be able to see list of candidates with their primary info, current state and score. Also there is an ability to change candidate state or delete candidate at all.

Screenshot 2020-08-25 at 4 30 44 PM
Screenshot 2020-08-25 at 4 31 01 PM

To Do list:

- [x] Refactoring;
- [ ] Add score percentages by field and score calculator on [Candidate View](https://join-react-test.herokuapp.com/applicant) to let candidate know how good is his application;
- [ ] Responsive layout.

### CODE REVIEW

File: src/components/Candidate/index.tsx [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+  const handleChangeNewState = (event: any) => {
+    setNewState(event.target.value);
+  };
+
+  const handleUpdateState = () => {
+    handleCloseDialog();
+    candidatesStore.updateCandidate(id, { state: newState });
+  };
+
+  useEffect(() => {
```
Comment (michowski) [inline src/components/Candidate/index.tsx:76]: How could you improve this approach?

File: src/components/Candidate/index.tsx [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    } else {
+      setProgressCircleClass(S.ProgressCircleOk);
+    }
+  }, [scoreColor, S]);
+
+  useEffect(() => {
+    setNewState(state);
+  }, [state]);
+
+  useEffect(() => {
```
Comment (michowski) [inline src/components/Candidate/index.tsx:90]: How could you improve this approach?

File: src/components/Candidate/index.tsx [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    if (scoreColor === 'error') {
+      setProgressCircleClass(S.ProgressCircleError);
+    } else if (scoreColor === 'warn') {
+      setProgressCircleClass(S.ProgressCircleWarning);
+    } else {
+      setProgressCircleClass(S.ProgressCircleOk);
+    }
+  }, [scoreColor, S]);
+
+  useEffect(() => {
```
Comment (michowski) [inline src/components/Candidate/index.tsx:86]: Do you see another way to accomplish this goal?

File: src/components/Candidate/index.tsx [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+
+const CandidateView = ({ candidate }: Props) => {
+  const S = useStyles();
+  const { candidatesStore } = stores;
+  const { id, fullName, email, avatar, state, appliedOn, score = {} } = candidate;
+  const { scorePercentage = 0, scoreTitle = '', scoreColor = '' } = score;
+  const [progressCircleClass, setProgressCircleClass] = useState(S.ProgressCircleError);
+  const [stateClass, setStateClass] = useState(S.State);
+  const currentState = states.find(item => item.label === state);
+  const availableStates = states.filter(item => !(currentState && item.step !== currentState.step && item.step !== currentState.step + 1));
```
Comment (michowski) [inline src/components/Candidate/index.tsx:38]: This line is pretty hard to follow. Could you improve it?

File: src/helpers/score.helper.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  { field: 'fullName', miltiplier: 1 },
+  { field: 'email', miltiplier: 1 },
+  { field: 'password', miltiplier: 1 },
+  { field: 'phone', miltiplier: 2 },
+  { field: 'avatar', miltiplier: 5 }
+];
+
+export const getCandidateScorePercentage = (candidate: any) => {
+  if (!candidate) return 0;
+  const total = candidateScoreMultiplier.reduce((sum, { miltiplier }) => (sum + miltiplier), 0);
```
Comment (michowski) [inline src/helpers/score.helper.ts:12]: We have a repetition of `candidateScoreMultiplier.reduce(...)`. Do you see a way to refactor and optimize it?

File: src/helpers/score.helper.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
+
+export const candidateScoreMultiplier = [
+  { field: 'fullName', miltiplier: 1 },
+  { field: 'email', miltiplier: 1 },
+  { field: 'password', miltiplier: 1 },
+  { field: 'phone', miltiplier: 2 },
+  { field: 'avatar', miltiplier: 5 }
+];
+
+export const getCandidateScorePercentage = (candidate: any) => {
```
Comment (michowski) [inline src/helpers/score.helper.ts:10]: Could you find a more suitable type for the `candidate` argument?

File: src/helpers/score.helper.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  const total = candidateScoreMultiplier.reduce((sum, { miltiplier }) => (sum + miltiplier), 0);
+  const passed = candidateScoreMultiplier.reduce((sum, { field, miltiplier }) => (
+    field && candidate[field] ? (sum + miltiplier) : sum
+  ), 0);
+  return Math.round((passed / total) * 100);
+};
+
+export const getCandidateScore = (candidate: any) => {
+  const scorePercentage = getCandidateScorePercentage(candidate);
+  let scoreColor = 'error';
```
Comment (michowski) [inline src/helpers/score.helper.ts:21]: How could you avoid the `let` keyword and replace it with `const`? Which one would you prefer and why?

File: src/helpers/validation.helper.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+  // create validation rules based on keys from store
+  Object.keys(rules).forEach(key => {
+    const value = rules[key];
+    validationRules[key] = defaultRules[value] ? defaultRules[value] : value;
+  });
+
+  const errors = validate(data, validationRules, options);
+
+  const result = errors && Object.keys(errors).map(key => {
```
Comment (michowski) [inline src/helpers/validation.helper.ts:25]: Do you see an issue with a `.map` function that always returns `true`?

File: src/stores/candidates.store.ts [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+      this.errors = error.errors;
+      this.loading = false;
+    }
+  })
+
+  deleteCandidate = flow(function* fetch(this: any, id: string) {
+    this.loading = true;
+    try {
+      const { item } = yield apiService.deleteCandidate(id);
+      const idx = this.candidates.findIndex((candidate: any) => candidate.id === id);
```
Comment (michowski) [inline src/stores/candidates.store.ts:139]: We have a few repetitions of:

```ts
const idx = this.candidates.findIndex((candidate: any) => candidate.id === id);
if (idx !== -1) { .... }
```

Would you be able to refactor it?

**Review by michowski — COMMENTED** [review]
Hi Julian, thank you very much for submitting your solution.

Would you like to answer a few comments that I have left? Please, feel free to push new commits to the PR if you come up with any improvement, be it to the parts I have commented or anything you feel like could be better.

Best,
Arek

**Comment by JulianKosach** [comment]
Hi Arek! Thanks for your comments.

I'll make a new commit for sure. I've known that some points should be
refactored (classes and colors in states and effects) but didn't have
enough time on Friday to do it. But at some points I didn't even notice
that they can be improved. So thank you a lot, I will push a new commit
ASAP.

Best,
Julian.

пн, 24 серп. 2020 о 19:00 Arek Michowski <notifications@github.com> пише:

File: src/stores/candidates.store.ts (hunk 9 above)
Comment (JulianKosach) [reply src/stores/candidates.store.ts]: Added new class method `getCandidateById(id: string)`

File: src/helpers/validation.helper.ts (hunk 8 above)
Comment (JulianKosach) [reply src/helpers/validation.helper.ts]: Yes ,there is no reason to use `[].map`. Replaced with `forEach`.

File: src/helpers/score.helper.ts (hunk 7 above)
Comment (JulianKosach) [reply src/helpers/score.helper.ts]: Replaced with `const score = { ... };`. Also refactored conditions a bit.

File: src/helpers/score.helper.ts (hunk 6 above)
Comment (JulianKosach) [reply src/helpers/score.helper.ts]: Yes. Replaced with Type `Candidate` from 'types/Candidate';

File: src/helpers/score.helper.ts (hunk 5 above)
Comment (JulianKosach) [reply src/helpers/score.helper.ts]: Yes, replaced with one `forEach` method call.

File: src/components/Candidate/index.tsx (hunk 4 above)
Comment (JulianKosach) [reply src/components/Candidate/index.tsx]: Yes, refactored.

File: src/components/Candidate/index.tsx (hunk 3 above)
Comment (JulianKosach) [reply src/components/Candidate/index.tsx]: Actually, there is no reason to use this useEffect, so it removed now.

File: src/components/Candidate/index.tsx (hunk 2 above)
Comment (JulianKosach) [reply src/components/Candidate/index.tsx]: Replaced with Material UI Class name branch.

File: src/components/Candidate/index.tsx (hunk 1 above)
Comment (JulianKosach) [reply src/components/Candidate/index.tsx]: Replaced with Material UI Class name branch.

**Comment by JulianKosach** [comment]
Hi Arek @michowski !

I've did my best to improve my PR. Please take a look. 
I will be waiting for your feedback.

Best,
Julian.
